[BEAM-3773][SQL] Add EnumerableConverter for JDBC support

apilloud opened the pull request:
This adds a converter from BeamRelNode to EnumerableRel. The calcite JDBC engine can execute any plan with a root node of the EnumerableRel type.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [X] Write a pull request description that is detailed enough to understand:
   - [X] What the pull request does
   - [X] Why it does it
   - [X] How it does it
   - [X] Why this approach
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

apilloud commented:
R: @kennknowles 
cc: @akedin

apilloud commented:
run java precommit

apilloud commented:
run java precommit

apilloud commented:
run java precommit

kennknowles commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java:
```diff
@@ -0,0 +1,160 @@
+  }
+
+  public static Enumerable<Object> toEnumerable(BeamRelNode node) {
+    PipelineOptions options = PipelineOptionsFactory.create();
```
Just checking - since `Collector` only works on the `DirectRunner` it seems fine to hardcode it here. But are the options specified elsewhere when using SQL so here it would just be validation that the configuration is supported?

apilloud replied:
Pipelines that aren't supported should get a `NullPointerException` when the `Collector` tries to process an element. I considered adding a check but decided it could wait because this is the only public path in. Is the user able to change the runner without the ability to set the `PipelineOptions`?

kennknowles replied:
I'd rather get the options plumbed deliberately here. Assuming decent testing, it won't get left forever, but it does leave the code in a sort of weird stage. Does it add a ton of scope to plumb them?

apilloud replied:
Plumbing the `PipelineOptions` into `BeamEnumerableConverter` is not easy as they will have to take a trip through the calcite JDBC interface on the way here (possibly on the jdbc connection string). I can add them as an argument to `toEnumerable`, but that will just move this line up to `implement`.

kennknowles commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java:
```diff
@@ -0,0 +1,160 @@
+  }
+
+  private static class Collector extends DoFn<Row, Void> {
+    // This will only work on the direct runner.
```
It seems like there are two routes that are interesting here: (1) expose the ability to observe PCollection contents in the DirectRunner. This used to be the case when it only supported bounded collections. It is a different situation now, as the contents are never materialized. But that could change.

The other thing that would potentially make it cross runner, and what Scio does, is to write it to a sink. It could be TextIO writing to the globally-required tempLocation. We should definitely learn from Scio either way.

apilloud replied:
I agree on writing to a sink. I was considering a version that sinks to BigQuery and read back from there. There are many options with different tradeoffs, we should consider them as followups.

kennknowles replied:
Yea, I think we probably want to be able to set options to control that if/when we do implement.

apilloud commented:
run java precommit

apilloud commented:
run java precommit

apilloud commented:
run java precommit

apilloud commented:
run java precommit

akedin commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java:
```diff
@@ -0,0 +1,162 @@
+
+    Collector.globalValues.put(id, values);
+    run(options, node, new Collector());
+    Collector.globalValues.remove(id);
```
Is it necessary to keep a static state? Cannot this be an instance field of the `Collector`?

apilloud replied:
The `Collector` is serialized by beam and a copy is passed to the worker. As a result each worker ends up with a copy of the collector, so the original queue won't contain any values added by the worker. The static state is circumventing Beam's requirement that there be no global state. This only works in the direct runner because all workers are in the same process.

kennknowles replied:
It seems like a pattern of concern that this only works on the direct runner but maybe is positioned in the codebase in a way that makes us think it should work more broadly. It could be solved just by really clear signaling and validation, perhaps, and maybe TODO JIRAs? I'm looking for options to make everyone happy and unblock your other work.

apilloud replied:
I'm explicitly violating the contract here, but being able to collect results in memory seems like a useful testing pattern for the direct runner. Is this something that might be able to live in a util folder inside the direct runner?

akedin commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java:
```diff
@@ -0,0 +1,162 @@
+/**
+ * BeamRelNode to replace a {@code Enumerable} node.
+ */
+public class BeamEnumerableConverter extends ConverterImpl implements EnumerableRel {
```
Can you document the behavior and how this class is used? And maybe an integration test of some kind?

apilloud replied:
A integration test would be good, I'll get one added. I started down the unit test path last week and ended up deleting all the code that was unit testable when I found a library that implemented it.

akedin commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamEnumerableConverter.java:
```diff
@@ -0,0 +1,162 @@
+        .apply(node.toPTransform())
+        .apply(ParDo.of(doFn));
+    PipelineResult result = pipeline.run();
+    result.waitUntilFinish();
```
Does this work in general case? E.g. what happens when the input is unbounded? My understanding is that `Enumerables` can be unbounded as well, and JDBC supports paginated unbounded results as well, is this going to work?

apilloud replied:
This does not work in the general case, it only works on pipelines that eventually finish. I wanted to get something in for my DDL work, we can make this better in a followup.

apilloud commented:
Fixed the issues from the comments and added tests at the `toEnumerable` layer. There are still some missing pieces for a functional calcite JDBC test, so that will have to come once those features are added.
